react-backend/client/src/cb-admin/pages/VisitorDetails.js

```diff
@@ -0,0 +1,225 @@
+  flex: ${props => props.flex || '1'};
+  height: 100%;
+`;
+
+const Form = Fm.extend`
+  width: 100%;
+`;
+
+const Row = FlexContainerRow.extend`
+  align-content: center;
+  align-items: flex-start;
+  flex: 3;
+`;
+
+const HyperLink = Link.extend`
+  flex: ${props => props.flex || '1'};
+`;
+
+const FormSection = styled.section`
+  width: 20%;
+`;
+
+
+const keyMap = {
+  visit_id: null,
+  visitor_id: 'Visitor ID',
+  gender: 'Gender',
+  yob: 'Year of birth',
+  activity: 'Activities',
+  visit_date: 'Date of visit',
+};
+
+const colToState = invertObj(keyMap);
+
+const columns = Object.values(keyMap).filter(Boolean);
+
+const sortOptions = [{ key: '0', value: '' }].concat(columns.map((col, i) => ({ key: `${i + 1}`, value: col })));
+
+const genderOptions = [
+  { key: '0', value: '' },
+  { key: '1', value: 'Male' },
+  { key: '2', value: 'Female' },
+  { key: '3', value: 'Prefer not to say' },
+];
+
+const ageOptions = [
+  { key: '0', value: '' },
+  { key: '1', value: '0-17' },
+  { key: '2', value: '18-35' },
+  { key: '3', value: '35-50' },
+  { key: '4', value: '51-69' },
+  { key: '5', value: '70+' },
+];
+
+
+const sortItems = (items, order, field) =>
+  sortBy(id => items[id][field], order);
+
+const currentYear = () => (new Date()).getFullYear();
+
+export default class VisitorDetailsPage extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      visits: {
+        items: {},
+        order: [],
+      },
+      sort: '',
+      genderFilter: '',
+      ageFilter: '',
+      errors: {},
+    };
+  }
+
+  componentDidMount() {
+    const cbAdminToken = this.props.auth;
+
+    Visitors.get(cbAdminToken, { withVisits: true })
+      .then((res) => {
+        this.props.updateAdminToken(res.headers.authorization);
+        const visits = res.data.result;
+
+        const order = visits.map(v => v.visit_id);
+        const items = visits.reduce((acc, v) => {
+          acc[v.visit_id] = { ...v, display: true };
+          return acc;
+        }, {});
+
+        this.setState(assocPath(['visits'], { order, items }));
+      })
+      .catch(error => console.log(error));
+  }
+
+  onChange = e =>
+    this.setState({ [e.target.name]: e.target.value }, this.update)
+
+  update = () => {
+    const { sort, genderFilter, ageFilter, visits: { items, order } } = this.state;
+    let newOrder = [...order];
+
+    if (sort) {
+      const by = sort && (colToState[sort] || 'visit_id');
+      newOrder = sortItems(items, order, by);
+    }
+
+    const newItems = map((item) => {
+      const displayForGender = genderFilter ? item.gender === genderFilter.toLowerCase() : true;
+      let displayForAge = true;
+
+      if (ageFilter) {
+        const age = currentYear() - item.yob;
+        const match = ageFilter.match(/(\d{1,2})[-+](\d{0,2})/);
+
+        if (match) {
+          const [_, min, max] = match; // eslint-disable-line
+          displayForAge = age > +min && age < (+max || Infinity);
```

Comment:
What's this doing?

Reply:
Setting a flag that determines whether the item in question has `yob` field that is compatible with the age filter.

Age filter is specified as `X-Y` or `Z+`, so I extract the numbers via regex, then do the comparison to set the flag.

Frankly it's a little slapdash, there's definitely higher order concepts in there, but it'll do for now.